Refactor script to be more user-friendly and maintainable

figure002 opened the pull request:
Refactor script to be more user-friendly and maintainable

This commit refactors the script to improve readability, usability, and
maintainability.

- The code has been broken down into smaller, more focused functions to
  improve readability.
- Code now follows PEP8. Run `make lint` to check the code.
- An argument parser has been added, allowing users to configure the
  script via the command line instead of modifying the code directly.
- The logging module is now used for better message handling.
- Type annotations have been added throughout the code to enable static
  type checking.

And manage the project with `uv`, an extremely fast Python package and
project manager. Simply run `uv sync` to create the project environment
with all dependencies.

figure002 committed:
Refactor script to be more user-friendly and maintainable

This commit refactors the script to improve readability, usability, and
maintainability.

- The code has been broken down into smaller, more focused functions to
  improve readability.
- Code now follows PEP8. Run `make lint` to check the code.
- An argument parser has been added, allowing users to configure the
  script via the command line instead of modifying the code directly.
- The logging module is now used for better message handling.
- Type annotations have been added throughout the code to enable static
  type checking.

And manage the project with `uv`, an extremely fast Python package and
project manager. Simply run `uv sync` to create the project environment
with all dependencies.

figure002 commented:
@Jerrythafast This pull request also renames the script to `transip-dyndns.py`, which is not great for reviewing and Git history/Git blame. If you agree with this change, you could first rename the file to `transip-dyndns.py` yourself, and then I will rebase this pull request, so that it's easier for you to review.

Jerrythafast commented:
Somehow this PR completely evaded my attention! Apologies, I don't like open source projects to be run like this, yet here we are.

I need a bit more to have a look at this, though I must say at first glance it looks quite tidy. Sorry again for not responding in 8 months.

figure002 commented:
@Jerrythafast No worries Jerry, I completely understand. Sometimes life gets in the way, or you just miss the notification. It happens to me all the time.

figure002 commented:
@Jerrythafast I have rebased onto the `master` branch. I've also renamed the script back to `dyndns.py` so that it's easier for you to review. I also switched to `uv` as the Python package and project manager. See the `REAME` for details.

figure002 commented:
@Jerrythafast Do you have time to review?

Jerrythafast commented on dyndns.py:
```diff
@@ -1,176 +1,347 @@
+
+import argparse
+import base64
+import ipaddress
```
Could also add `AddressValueError` to the `from ipaddress import ...` line.
(Also, how did I not notice the existence of this package?!)

Jerrythafast commented on dyndns.py:
```diff
@@ -1,176 +1,347 @@
+from cryptography.hazmat.primitives.serialization import load_pem_private_key
+
+
+LOG_FORMAT = "%(levelname)s %(message)s"
```
The `LOG_FORMAT` no longer includes the date/time, which I found really useful.
I'd prefer the following format, for a clean look including date and time:
```suggestion
LOG_FORMAT = "%(asctime)s  %(levelname)-7s  %(message)s"
```

Jerrythafast commented on dyndns.py:
```diff
@@ -1,176 +1,347 @@
+            logger.info(f"Got invalid response from {url}, skipping")
+
+    if last_exc is not None:
+        logger.exception(
```
`logger.exception()` prints a stack trace, which breaks the neat log format when any requests fail. I'd prefer the previous look, which could be replicated with:
```python
logger.error(f"Failed to determine current IPv{version} address. {type(last_exc).__name__}: {last_exc}")
```
Or with the `format_exception()` function reinstated: 
```python
logger.error(f"Failed to determine current IPv{version} address. {format_exception(ex)}")
```

Jerrythafast commented on dyndns.py:
```diff
@@ -1,176 +1,347 @@
+    parser.add_argument(
+        "-k",
+        "--private-key",
+        type=argparse.FileType("r"),
```
`argparse.FileType` is now deprecated (it wasn't when you wrote this). I think it's best to just use the default type (`str`) to open/close the file manually using a `with` statement.

Jerrythafast commented on dyndns.py:
```diff
@@ -1,176 +1,347 @@
+        parser.error(str(exc))
+
+    if args.ipv4 == IPV4_UNSPECIFIED:
+        logger.info("Determining IPv4 address...")
```
I'd prefer the "Determining IPv# address..." and "Your IPv# address: {}" messages to be logged at `DEBUG` level to reduce log spam. (A successful run of the current script version only logs "No changes are required for {domain}".) Maybe the `DEBUG` log level could then be enabled with a new `--debug` or `--verbose` option.

Jerrythafast commented on pyproject.toml:
```diff
@@ -0,0 +1,18 @@
+version = "0.1.0"
+description = "Dynamic DNS updating script for TransIP API"
+readme = "README.md"
+requires-python = ">=3.13"
```
I'm successfully running this on much older versions, I think the `README.md` is correct in that this runs on 3.6+.

Jerrythafast commented on README.md:
```diff
@@ -1,38 +1,97 @@
+    ip addr show eth0 | grep -Poh '(?<=inet6\s)[0-9a-f:]+(?=/64)' | grep -vm 1 '^fe80'
+
+
+## Development
```
I am myself not using `uv` nor `make`, so all of that (basically anything other than the `dyndns.py` and `README.md` and perhaps the basics of `pyproject.toml`) are not really of use to me. I'd prefer to keep the repository "small and simple" without it if possible.

Jerrythafast commented on README.md:
```diff
@@ -1,38 +1,97 @@
+3. Make the script executable: `chmod +x dyndns.py`
+4. See `./dyndns.py --help` for usage instructions.
+5. Execute the `dyndns.py` script on the command-line whenever you want
+   the DNS to be automatically updated. You can use `cron` to do this
```
I was thinking, now that we have the key in a file, on Systemd-enabled systems an even neater approach exists. One could make a Systemd `timer` unit that uses the `LoadCredential=` setting. The key file could then be made readable by absolutely no-one (other than `root`; i.e., `chmod 000`) and it gets securely exposed to the dyndns script.

Jerrythafast reviewed (changes requested):
First of all, apologies again for the insane delay in reviewing your work.

Your changes are definitely a big improvement in code quality throughout the script. I've only got a couple of small comments here and there, but overall I absolutely agree with the general way things have changed.

I'm just a bit uncertain about the packaging/tooling getting into the repository as well, especially considering, in essence, it is just a single self-contained script. I think that, for such a simple project, we can do without `uv` and the `Makefile`, leaving the choice about which tools to use for dependency management and deployment to the users of the script.

Thanks a lot!!
